Adds hauntium as a reagent with new fun uses.

dieamond13 opened the pull request:
## About The Pull Request
Adds the material hauntium as a reagent, along with bringing some unique effects with the reagent. 

There are a few new ways to get it other than just the camera obscura as before, such as grinding certain ghostly things.

Unique effects it brings:

- It works in metalgen (I believe the metalgen recipe is still broken as of making this) to make hauntium infused metalgen.

https://github.com/tgstation/tgstation/assets/58376695/121ca6b2-00b7-4203-bc6a-48976b1af802


- Any object touched by the chem will temporarily have the haunted effect, letting it teleport around randomly and even attack people.


https://github.com/tgstation/tgstation/assets/58376695/103aeee5-ed3e-4fdd-85e4-ac7338823c46


- If you put it in your body, it hurts your "soul" (soul being heart and mood)
![image](https://github.com/tgstation/tgstation/assets/58376695/34182a4c-face-4fe7-b170-a7b10122a14d)
...unless you are morbid or undead, in which it gives a slight mood boost and acts like an addiction-less drug
## Why It's Good For The Game
Hauntium previously was super niche and not very useful other than making funny haunted toolboxes from it. This expands it's usage and makes it a lot more interesting. I also know a good few people who have also wished this was available to use in metalgen and now you can!
## Changelog
:cl:
add: Lets you grind things into a hauntium reagent which works similarly to the solid form but is versatile and has some unique effects.

/:cl:

dieamond13 committed:
Adds hauntium reagent

Adds hauntium as a reagent with unique interactions.

dieamond13 committed:
Adds hauntium as a reagent

Adds hauntium as a reagent and some unique effects with it.

necromanceranne commented:
Probably shouldn't have negative effects on ingest for the undead and morbid. Haunted objects think they're cool.

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -2891,3 +2891,29 @@
+
+/datum/reagent/hauntium/expose_obj(obj/exposed_obj, volume) //gives 15 seconds of haunting effect for every unit of it that touches an object
+	. = ..()
+	exposed_obj.AddElement(/datum/element/haunted, 0)
+	addtimer(CALLBACK(exposed_obj, PROC_REF(_RemoveElement), list(/datum/element/haunted, 0)), volume * time_multiplier, TIMER_UNIQUE|TIMER_OVERRIDE)
```
This feels a bit icky. Beside the callback's proc having an underscore prefix, which usually means it shouldn't be deliberately called without using a wrapper of some sort, this is likely going to conflict with the hauntium material and revenant ability, which also add and remove the same element, at some point.

One solution to the problem would be adding an `atom/movable` proc that checks if the movable already has a haunted trait, add the element if that statement is false, then adds the trait. As well as  another that does the opposite.

More or less like this:
```
/atom/movable/proc/make_haunted(source, color)
	if(!HAS_TRAIT(src, TRAIT_HAUNTED))
		AddElement(/datum/element/haunted, color)
	ADD_TRAIT(src, TRAIT_HAUNTED, source)

/atom/movable/proc/remove_haunted(source)
	REMOVE_TRAIT(src, TRAIT_HAUNTED, source)
	if(!HAS_TRAIT(src, TRAIT_HAUNTED))
		RemoveElement(/datum/element/haunted)
```

dieamond13 replied:
I just took it from the gravitums code but instead of no gravity it gives it haunting, I can try and do something more like that though

dieamond13 replied:
if I were to put those procs in the haunted element page, datums/elements/haunted.dm, how would I call them from reagents / anywhere else I may want to use it?

Ghommie replied:
As long as it's an atom/movable proc, you can basically call `exposed_obj.make_haunted(trait_source_define_goes_here, color_string_goes_here)` and `exposed_obj.remove_haunted(trait_source_define_goes_here)` and you should be set.

dieamond13 commented:
> Probably shouldn't have negative effects on ingest for the undead and morbid. Haunted objects think they're cool.

Ah I didnt even know that was a thing, yeah I'll add that in.

dieamond13 committed:
You are no longer negatively effected by hauntium if you are undead or morbid.

If you are morbid or undead, instead of getting a strong negative moodlet you get a small positive moodlet, and as well as the reagent acting like a drug instead.

nianjiilical commented:
mfw my heavily mutated botany strange seeds with 16 negative reagents and/or high power alcohol types suddenly become infested with ghosts and start beating the shit out of me

(this is a good thing)

Vincent983 commented:
Can't wait to make a grenade that makes a shit ton of meat and sprays it all with hauntium so they slap people.

dieamond13 committed:
Adds morbid/undead interactions and new haunted traits

Reworks how the haunted element is generally applied and removed in the code, and adds special interactions for the morbid/undead.

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -2900,20 +2900,36 @@
 	. = ..()
-	exposed_obj.AddElement(/datum/element/haunted, 0)
-	addtimer(CALLBACK(exposed_obj, PROC_REF(_RemoveElement), list(/datum/element/haunted, 0)), volume * time_multiplier, TIMER_UNIQUE|TIMER_OVERRIDE)
+	exposed_obj.make_haunted(HAUNTIUM_REAGENT_TRAIT, "#f8f8ff")
+	haunt_timer = addtimer(CALLBACK(exposed_obj, TYPE_PROC_REF(/atom/movable/, remove_haunted), HAUNTIUM_REAGENT_TRAIT), volume * time_multiplier, TIMER_UNIQUE|TIMER_STOPPABLE)
+	if(haunt_timer) //if the timer for haunt removal is going and its splashed again, reset the timer.
+		deltimer(haunt_timer) //this looks kind of scuffed but every other way I tried it it just wouldn't work properly
+		haunt_timer = addtimer(CALLBACK(exposed_obj, TYPE_PROC_REF(/atom/movable/, remove_haunted), HAUNTIUM_REAGENT_TRAIT), volume * time_multiplier, TIMER_UNIQUE|TIMER_STOPPABLE)
```
I'm a tad perplexed. While it may work on a one to one basis, it won't if the same reagent is splashed on several objects, or when several instances of the reagent splashed on the same object. At this point, let's keep it simple, even if it means hauntium smoke and foam won't be as strong. Also, I don't believe there's a need for a `time_multiplier` variable.
```suggestion
	if(HAS_TRAIT_FROM(exposed_obj, TRAIT_HAUNTED, HAUNTIUM_REAGENT_TRAIT)
		return
	exposed_obj.make_haunted(HAUNTIUM_REAGENT_TRAIT, "#f8f8ff")
	addtimer(CALLBACK(exposed_obj, TYPE_PROC_REF(/atom/movable/, remove_haunted), HAUNTIUM_REAGENT_TRAIT), volume * 20 SECONDS)
```

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -2900,20 +2900,36 @@
 	material = /datum/material/hauntium
 	ph = 10
-	var/time_multiplier = 15 SECONDS
+	var/time_multiplier = 20 SECONDS
+	var/haunt_timer
```
```suggestion
```

dieamond13 committed:
Simplifies hauntium reagents interaction with items

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

dieamond13 committed:
Removes time_multiplier and haunt timer variables

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

dieamond13 committed:
Merge branch 'master' into hauntium-reagent

Ghommie commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -2891,3 +2891,42 @@
+
+/datum/reagent/hauntium/expose_obj(obj/exposed_obj, volume) //gives 15 seconds of haunting effect for every unit of it that touches an object
+	. = ..()
+	if(HAS_TRAIT_FROM(exposed_obj, TRAIT_HAUNTED, HAUNTIUM_REAGENT_TRAIT)
```
```suggestion
	if(HAS_TRAIT_FROM(exposed_obj, TRAIT_HAUNTED, HAUNTIUM_REAGENT_TRAIT))
```
Brackets, they get me everytime.

Ghommie committed:
Update code/modules/reagents/chemistry/reagents/other_reagents.dm
